### PR TITLE
Knuth-Morris-Pratt and Z-Algorithm

### PR DESCRIPTION
**Pull request by dontfollowmeimcrazy**
Implemented Knuth-Morris-Pratt algorithm for string searching, based on Z-Algorithm for the pattern preprocessing stage. 
Added a string search algorithm just by using Z-Algorithm (computing it on the concatenation pattern +"$"+text).
All algorithms are implemented after the pseudocode contained in "Algorithms on Strings, Trees and Sequences" by Dan Gusfield.

TODO: ReadMes

### CODE REVIEW

**Comment by dontfollowmeimcrazy** [comment]
Done with the sources and the READMEs. Hope that everything is good.

**Comment by chris-pilcher** [comment]
Thanks heaps @dontfollowmeimcrazy 👍 I'll try to get this reviewed/merged this weekend.

**Comment by kelvinlauKL** [comment]
@chris-pilcher Could you follow up with this?

File: Knuth-Morris-Pratt/KnuthMorrisPratt.playground/Contents.swift [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+    let pattern = Array(ptnr.characters)
+    let patternLength: Int = pattern.count
+
+    guard patternLength > 0 else {
+        return nil
+    }
+
+    var zeta: [Int] = [Int](repeating: 0, count: patternLength)
+
+    var left: Int = 0
```
Comment (kelvinlauKL) [inline Knuth-Morris-Pratt/KnuthMorrisPratt.playground/Contents.swift]: As per our style guide, leave out the explicit typing as Swift can infer the type from the assigned value.

```
// not preferred
var left: Int = 0

// preferred
var left = 0
```

File: Knuth-Morris-Pratt/KnuthMorrisPratt.playground/Contents.swift [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+    var textIndex: Int = 0
+    var patternIndex: Int = 0
+
+    for k in 1 ..< patternLength {
+        if k > right {
+            patternIndex = 0
+
+            while k + patternIndex < patternLength  &&
+                pattern[k + patternIndex] == pattern[patternIndex] {
+                patternIndex = patternIndex + 1
```
Comment (kelvinlauKL) [inline Knuth-Morris-Pratt/KnuthMorrisPratt.playground/Contents.swift]: Could be more succinctly written as

```
patternIndex += 1
```

File: Knuth-Morris-Pratt/KnuthMorrisPratt.playground/Contents.swift [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+        let patternLength: Int = pattern.count
+
+        guard patternLength > 0 else {
+            return nil
+        }
+
+        var suffixPrefix: [Int] = [Int](repeating: 0, count: patternLength)
+        var textIndex: Int = 0
+        var patternIndex: Int = 0
+        var indexes: [Int] = [Int]()
```
Comment (kelvinlauKL) [inline Knuth-Morris-Pratt/KnuthMorrisPratt.playground/Contents.swift]:
```
// not preferred
var indexes = [Int]()

// preferred
var indexes: [Int] = []
```

**Comment by dontfollowmeimcrazy** [comment]
Thank you 👍

**Pull request merged**